lib/addons/ssm-agent/index.ts

```diff
@@ -0,0 +1,87 @@
+import { KubernetesManifest } from "@aws-cdk/aws-eks";
+import { ManagedPolicy } from "@aws-cdk/aws-iam";
+import { ClusterAddOn, ClusterInfo } from "../../stacks/cluster-types";
+
+export class SSMAgentAddOn implements ClusterAddOn {
+    deploy(clusterInfo: ClusterInfo): void {
+        const cluster = clusterInfo.cluster;
+        console.assert(clusterInfo.nodeGroup || clusterInfo.autoscalingGroup, "SSMAgentAddon can only be used with EKS EC2 at the moment. "
+            + "If using customer cluster provider, make sure you return the node group");
+
+        // Setup managed policy.
+        const nodeGroup = clusterInfo.nodeGroup || clusterInfo.autoscalingGroup;
+        // Add AWS Managed Policy for SSM
+        nodeGroup!.role.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore'));
+
+        // Apply manifest.
+        // See APG Pattern https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/install-ssm-agent-on-amazon-eks-worker-nodes-by-using-kubernetes-daemonset.html
+        const appLabel = { app: "ssm-installer"};
+
+        const daemonSet = {
+            apiVersion: "apps/v1",
+            kind: "DaemonSet",
+            metadata: {
+                name: "ssm-installer",
+            },
+            spec: {
+                selector: { matchLabels: appLabel },
+                template: {
+                    metadata: { labels: appLabel },
+                    spec: {
+                        containers: [
+                            {
+                                image: "amazonlinux",
+                                imagePullPolicy: "Always",
+                                name: "ssm",
+                                command: ["/bin/bash"],
+                                args: [
+                                    "-c",
+                                    "echo '* * * * * root yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm & rm -rf /etc/cron.d/ssmstart' > /etc/cron.d/ssmstart"
```

Comment:
OMG, installs on every run? Isn't there an image with SSM pre-installed? This mutation won't work for any compliance programs, i am afraid. I am curious when eksctl is used with ssm = true, are they doing the same? (I realize that this is per documentation).

Reply:
As per https://github.com/awslabs/amazon-eks-ami/issues/404 and https://github.com/awslabs/amazon-eks-ami/issues/127, SSM agent is not considered required for EKS and hence not added in the AMI.

Reply:
I was thinking more along the lines of an AWS authored container image with SSM preinstalled. It would be nice to have it in AMI, but we also custom AMIs which won't have it.

Reply:
A few things before we close it out:

1. Let's use https://github.com/jicowan/ssm-agent to build the image and upload it to a public ECR.
2. We need public ECR repo for a number of things including bootstrapping so let's create a task for that (@kcoleman731 wdyt?)
3. Let's upvote the following on GitHub: https://github.com/aws/amazon-ssm-agent/issues/135 (@parkand1 @youngjeong46 please join in).

Reply:
@shapirov103 added $1 from above comment. Will need $2 (cc: @kcoleman731)

Reply:
Just making sure, as it stands it will work by pulling public image from your repository, correct? I am going to merge it right away to test if that is the case.

Reply:
That's correct. Any one can docker pull that image.